添加C++新标准012:enable_If (#27)
ttttLarva/Larva: docs/learn_cpp/012_enable_if.md

# 新标准：enable_if
:earth_asia: **Bilibili视频传送门：**[C++新标准012  enable_if](https://www.bilibili.com/video/BV1Tv4y1N73A?spm_id_from=333.999.0.0&vd_source=0184d13a7c21515c19c1cdb9230751c8) :earth_asia:

> [上一期视频](https://www.bilibili.com/video/BV1yr4y1t7qo?spm_id_from=333.999.0.0&vd_source=0184d13a7c21515c19c1cdb9230751c8)我们介绍了SFINAE原则，在此基础上，本篇文章将继续介绍了enable_if新标准，为后续深入学习模板元编程打下基础。

Reviewer: ```suggestion
[上一期视频](https://www.bilibili.com/video/BV1yr4y1t7qo?spm_id_from=333.999.0.0&vd_source=0184d13a7c21515c19c1cdb9230751c8) 我们介绍了 SFINAE 原则，在此基础上，本篇文章将继续介绍了 `enable_if` 新标准，为后续深入学习模板元编程打下基础。
```

- 空格问题还是需要自己重新自己检查下
- 代码的部分，用反斜杠扩一下

Reviewer: done

本文的结构如下：

> - 为什么要使用enable_if新标准？
> - enable_if&is_same的使用
> - enable_if&is_same内部的实现原理

Reviewer: ```suggestion
- 为什么要使用 `enable_if` 新标准？
- `enable_if` 与 `is_same` 的使用
- `enable_if` 与 `is_same` 内部的实现原理
```

同上（空格、代码）。以及使用 `&` 容易引起误解是运算符，不如直接使用“与”。视频和文字表达毕竟有区别，不要照搬，要思考哪种方式更好，为什么好

Reviewer: done

> ## 为什么要使用enable_if新标准？

Reviewer: ```suggestion
## 为什么要使用 enable_if 新标准？
```

空格问题，不再提醒了，请自己检查

Author: done

> C++模板函数重载依赖于 SFINAE (substitution-failure-is-not-an-error) 原则，即替换失败不认为是错误，编译器将会继续寻找合适的重载函数；以上一期视频中返回长度的函数为例，该函数针对STL中的vector容器来实现（容器中有size()成员函数）

Reviewer: ```suggestion
C++模板函数重载依赖于 SFINAE (substitution-failure-is-not-an-error) 原则，即替换失败不认为是错误，编译器将会继续寻找合适的重载函数；以上一期视频中返回长度的函数为例，该函数针对 STL 中的 `vector` 容器来实现（容器中有`size()` 成员函数）
```

代码，不再提示了，请自己检查

Author: done

```c++

> #SFINAE原则的应用

Reviewer: ```suggestion
## SFINAE 原则的应用
```

一级标题是整个文章的标题，这里是不是想用的是二级标题

class CMyclass
{
public:
    using size_type = unsigned int;
};

template <typename T>
decltype(T().size(), typename T::size_type()) len(T const &t) //该函数可以避免匹配到CMyclass类
{
    return t.size();
}
unsigned len(...)
{
    return 0;
}

int main()
{
    vector<int>vec={1,2,3};
    cout<<len(vec)<<endl;  //调用针对stl库容器的len函数
    cout << len(CMyclass()) << endl; //调用len(...)函数
}
```
上期视频已经介绍了重载函数如何实现避免匹配到CMyclass类，这种编程技巧需要程序员根据特例类不存在的方法制因此造“错误”,并不通用；

因此C++11采用了具有通用性的enable_if新标准来实现SFINAE；我们根据上期视频给出的测试程序，将重载函数的形式改写成使用enable_if的形式
```c++
#enable_if的应用
class CMyclass
{
public:
    using size_type = unsigned int;
};

// template <typename T>   //上期视频采用的编程技巧
// decltype(T().size(), typename T::size_type()) len(T const &t)
// {
//     return t.size();

> // }

Reviewer: 形成一篇技术博文，不是说非得一字不差地照搬视频中的内容，文字的表达和视频的表达方式可以有区别。

比如这里直接带上注释的代码，别人可能会觉得莫名其妙。

而变通的话，应该可以分两端，比如，“之前我们是这样些的 XXXX”，“现在可以改成这样 XXXX“

这个过程，对于当前做的这个 PR，是提高 PR 的质量。

对于更长远（而不直接）的影响，是锻炼如何和人异步、基于文字做技术交流。需要自己检查和改进下

Reviewer: 这个 comment 之后的内容我还没有看。可以先整体自己认真 refine 后，我再继续看。

Author: 收到 会针对反馈意见进行修改

# 新标准：enable_if
:earth_asia: **Bilibili视频传送门：**[C++新标准012  enable_if](https://www.bilibili.com/video/BV1Tv4y1N73A?spm_id_from=333.999.0.0&vd_source=0184d13a7c21515c19c1cdb9230751c8) :earth_asia:

[上一期视频](https://www.bilibili.com/video/BV1yr4y1t7qo?spm_id_from=333.999.0.0&vd_source=0184d13a7c21515c19c1cdb9230751c8)我们介绍了SFINAE原则，在此基础上，本篇文章将继续介绍了enable_if新标准，为后续深入学习模板元编程打下基础。

本文的结构如下：
- 为什么要使用`enable_if`？
- `enable_if`与`is_same`的使用
- `enable_if`与`is_same`内部的实现原理

## 为什么要使用 enable_if ？

> C++模板函数重载依赖于 SFINAE (substitution-failure-is-not-an-error) 原则，即替换失败不认为是错误，编译器将会继续寻找合适的重载函数；以返回长度的函数为例，该函数针对 STL 中的`vector`容器来实现（容器中有`size()`成员函数）

Reviewer: ```suggestion
## 为什么要使用 enable_if ？

C++模板函数重载依赖于 SFINAE (substitution-failure-is-not-an-error) 原则，即替换失败不认为是错误，编译器将会继续寻找合适的重载函数；以返回长度的函数为例，该函数针对 STL 中的`vector`容器来实现（容器中有`size()`成员函数）
```

虽然在  mkdocs 里没有实际影响，不过 markdown 的标准，标题和正文之间要有个空行（可能其它场景下会影响显示效果）

Author: done

```c++
class CMyclass
{
public:
    using size_type = unsigned int;
};

template <typename T>
decltype(T().size(), typename T::size_type()) len(T const &t) //该函数可以避免匹配到CMyclass类
{
    return t.size();
}
unsigned len(...)
{
    return 0;
}

int main()
{
    vector<int>vec={1,2,3};
    cout<<len(vec)<<endl;  //调用针对stl库容器的len函数
    cout << len(CMyclass()) << endl; //调用len(...)函数
}
```

> `decltype(T().size(), typename T::size_type()) len(T const &t) `该函数声明通过`decltype`类型判别将不包含`size()`成员函数的特例类排除出去，同时逗号表达式返回逗号后面的参数`typename T::size_type()`，实现避免匹配到`CMyclass`类的功能；这种编程技巧晦涩难懂，需要程序员根据特例类不存在的方法制因此造“错误”,并不通用；

Reviewer: ```suggestion
`decltype(T().size(), typename T::size_type()) len(T const &t)` 该函数声明通过 `decltype` 类型判别将不包含 `size()` 成员函数的特例类排除出去，同时逗号表达式返回逗号后面的参数 `typename T::size_type()`，实现避免匹配到 `CMyclass` 类的功能；这种编程技巧晦涩难懂，需要程序员根据特例类不存在的方法制因此造“错误”,并不通用；
```

> 为了实现 SFINAE ，同时降低程序的繁琐度，C++11 采用了具有通用性的`enable_if`新标准代替这种技巧 ；我们使用`enable_if`的形式重写该函数

Reviewer: ```suggestion
为了实现 SFINAE ，同时降低程序的繁琐度，C++11 采用了具有通用性的 `enable_if` 新标准代替这种技巧 ；我们使用 `enable_if` 的形式重写该函数
```

template<bool B,typename T=void>struct enable_if
```
- 当传入的模板参数 `B` 为 `True` 的时候，`enable_if` 模板类的 `type` 就等于 `T` ；如果不传入模板参数 `T` ， `type` 也会默认为 `void` 
- 当传入的模板参数 `B` 为 `False` 的时候， `enable_if` 模板类就不再拥有 `type` 。

使用 `enable_if` 时可以让重载函数自由丢弃不需要匹配的特例类；下面给出实际应用的例子

```c++
#判断类型字节数是否大于4输出显示
template <typename T>
typename std::enable_if<(sizeof(T) <= 4)>::type show() //针对 `T` 的长度小于等于4的情况生效
{
    printf("size<=4\n");
}

template <typename T>
typename std::enable_if<(sizeof(T) > 4)>::type show()  //针对 `T` 的长度大于4的情况生效
{
    printf("size>4\n");
}

int main()
{
    show<double>();
    show<char>();
}
```
输出结果为
```c++
size>4
size<=4
```

### is_same 的使用

```c++
template<class T,class U>struct is_same
```
- 当模板参数 `T` 等于模板参数 `U` 时, `is_same<T,U>`模板类的成员 `value` 等于 `True`
- 当模板参数 `T` 不等于模板 `U` 时, `is_same<T,U>`模板类的成员 `value` 等于 `False`

`is_same` 经常与 `eanble_if` 搭配使用
```c++
template <typename T, typename T2 = typename enable_if<!is_same<T, CMyclass>::value>::type> //`enable_if`形式
typename T::size_type len(T const &t)
{
    return t.size();
}
```
`enable_if` 模板类的第一个参数经常用 `is_same` 的 `value` 来代替，结合起来对一些特例进行判断，实现 SFINAE

##  enable_if 与 is_same 内部的实现原理

了解这两个模板类的内部实现原理之前，需要了解 type traits 的概念

> ### 什么是  type traits?

Reviewer: ```suggestion
### 什么是 type traits?
```

多了个空格

Author: done

>  traits 是 c++ 模板编程中使用的一种技术，主要功能： 

Reviewer: ```suggestion
 traits 是 C++ 模板编程中使用的一种技术，主要功能： 
```

大小写注意下

Author: done

 把功能相同而参数不同的函数抽象出来，通过 traits 将不同的参数的相同属性提取出来，在函数中利用这些用 traits 提取的属性，使得**函数对不同的参数表现一致**。

### enable_if 源码解析

下面给出 `enable_if` 的源码
```c++
template<bool, typename _Tp = void>
struct enable_if
{ };

template<typename _Tp>
struct enable_if<true, _Tp>
{ 
     typedef _Tp type; 
};
```
- 首先定义了一个空的通用模板，不包含任何成员
- 当第一个模板参数值为 `True` 时，将模板特例化; `type` 的类型为第二个模板参数 `_Tp` 。

> ###  is_same 源码解析

Reviewer: ```suggestion
### is_same 源码解析
```

多了个空格

Author: done
